docs/proposals/20191106-machine-load-balancer.md

```diff
@@ -0,0 +1,121 @@
+  * The MLB depends on the CAPI Machine CRD'S but it could be extended to other machine like objects as well.
+* CAPI has no dependency on the MLB, it only waits until something sets a controlPlaneEndpoint 
+
+### Workflow
+
+
+![Figure 1](./images/machine-load-balancer/workflow.png)
+
+
+
+**1 - 4:** Controllers watch the resources they responsible for
+
+**5 - 7:** User or system creates all CRD's *concurrently*
+
+**8 - 9:**  Infra provider spins up infrastructure and then waits until the `spec.controlPlaneEndpoint` is set
+
+**10 -11:** The MLB provisions the load balancer and updates `status.endpoint`
+
+**12: ** The template operator picks up that `status.endpoint` has been updated and copies the value to `spec.controlPlaneEndpoint`
+
+**13+** CAPI infra providers pick up that the `controlPlaneEndpoint` has been set and continue provisioning 
+
+##   Types
+
+```go
+type MachineLoadBalancer struct {  
+  ObjectMeta  
+  TypeMeta    
+  Spec MachineLoadBalancerSpec    
+  Status MachineLoadBalancerStatus
+} 
+type MachineLoadBalancerSpec struct {   
+  Port MachineLoadBalancerPort     
+  // MachineSelector provides a mechanism to select multiple machines based on their labels        // allowing for load balancing across clusters and different network topologies.   
+  MachineSelector Selector    
+  // ProviderRef is a load balancer specific CRD used for configuring the underlying load                       // balancer and/or providing access credentials    
+  ProviderRef corev1.ObjectReference
+} 
+
+type MachineLoadBalancerPort struct {    
+  Port int    
+  MachinePort int    
+  Protocol string
+} 
+
+type MachineLoadBalancerStatus struct { 
+  // Endpoint is the DNS/IP endpoint for the load balancer (includes scheme and port)    	    
+  Endpoint string     
+  // Status provides the health of the load balancer    
+  Status string    
+  // Ready is true when there is at least 1 healthy member    
+  Ready bool
+}
+```
+#### Contract with CAPI
+The Machine Load Balancer is decoupled from CAPI such that CAPI and infra providers are not aware of the existence of the MachineLoadBalancer - However the following conditions are required:
+
+* Infra providers MUST NOT provision a load balancer if an external control plane endpoint is already specified
+* Machines MUST be labelled with:
+  * **cluster.x-k8s.io/control-plane**: "" if the machine is part of the control plane
+  * **cluster.x-k8s.io/cluster-name** the name of the cluster
+  * **cluster.x-k8s.io/control-plane/selected-for-upgrade**=true if the machine is queued for deletion and its connections should be drained, and then removed from a LB.
+* Machines SHOULD be labelled with:
+  * **failure-domain.beta.kubernetes.io/zone** if applicable
+  * **failure-domain.beta.kubernetes.io/region** if applicable
+
+
+## User Stories
+
+#### Support for existing infrastructure
+
```

Comment:
There should be consistent use of blank lines after headers. This is good, but many of the user stories below lack the blank lines. I won't list every one.